### PR TITLE
Set up framework for an `assert_equal` function

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

While `qml.equal` has been incredibly helpful in helping us compare operators, measurements, and tapes, it can sometimes be rather difficult to determine *why* two operations are not equal.

This new `assert_equal` function will raise and `AssertionError` with context, making it easier to debug failures occuring in tests or scripts.

**Description of the Change:**

This PR just sets up the framework by:

1) Allowing the `_equal` function to return either a bool or str.
2) Making `qml.equal` interpret a string as `False`
3) Adding `assert_equal` to raise an `AssertionError` from the output of `_equal`
4) Implementing the bool-> str conversion for `BasisRotation`

`BasisRotation` is as it an edge case where I can demonstrate what we want these functions to look like without taking away the core part of the work.

**Benefits:**

Easier debugging.

**Possible Drawbacks:**

More to maintain.

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -60,6 +60,8 @@
   `qml.devices.Device`, which follows the new device API.
   [(#5581)](https://github.com/PennyLaneAI/pennylane/pull/5581)
 
+* Sets up the framework for the development of an `assert_equal` function for testing operator comparison.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:63]:
```suggestion
* Sets up the framework for the development of an `assert_equal` function for testing operator comparison.
  [(#5623)](https://github.com/PennyLaneAI/pennylane/pull/5634)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -60,6 +60,9 @@
...
+  [(#5623)](https://github.com/PennyLaneAI/pennylane/pull/5634)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:64]:
```suggestion
  [(#5634)](https://github.com/PennyLaneAI/pennylane/pull/5634)
```

File: pennylane/ops/functions/equal.py [hunk 3]
```diff
@@ -163,14 +163,99 @@ def equal(
...
+            Not used for comparing ``MeasurementProcess``, ``Hamiltonian`` or ``Tensor`` objects.
+        rtol (float, optional): Relative tolerance for parameters. Not used for comparing ``MeasurementProcess``, ``Hamiltonian`` or ``Tensor`` objects.
+        atol (float, optional): Absolute tolerance for parameters. Not used for comparing ``MeasurementProcess``, ``Hamiltonian`` or ``Tensor`` objects.
+
+    Returns:
+        None
+
+    Raises:
+
+        AssertionError: An `AssertionError` is raised if the two operators are not equal.
```
Comment (trbromley) [inline pennylane/ops/functions/equal.py:204]:
```suggestion
        AssertionError: An ``AssertionError`` is raised if the two operators are not equal.
```

File: pennylane/ops/functions/equal.py [hunk 4]
```diff
@@ -163,14 +163,99 @@ def equal(
...
+    0.        -0.09983342j]
+    [0.        +0.j         0.99500417+0.j         0.        -0.09983342j
+    0.        +0.j        ]
+    [0.        +0.j         0.        -0.09983342j 0.99500417+0.j
+    0.        +0.j        ]
+    [0.        -0.09983342j 0.        +0.j         0.        +0.j
+    0.99500417+0.j        ]].
+    >>> mat3 = qml.numpy.array(0.3)
+    >>> op3 = qml.BasisRotation(wires=(0,1), unitary_matrix = mat3)
+    >>> assert_equal(op1, op3)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/functions/equal.py:238]: My idea was to put it in the docs and import it top level once it was finished.

File: tests/ops/functions/test_equal.py [hunk 5]
```diff
@@ -2062,16 +2073,24 @@ class TestBasisRotation:
     @pytest.mark.parametrize("op, other_op", [(op1, op3)])
     def test_different_tolerances_comparison(self, op, other_op):
         assert qml.equal(op, other_op, atol=1e-5)
+        assert_equal(op, other_op, atol=1e-5)
         assert qml.equal(op, other_op, rtol=0, atol=1e-9) is False
 
+        with pytest.raises(AssertionError, match="The hyperparameter unitary matrix is not equal"):
```
Comment (albi3ro) [inline tests/ops/functions/test_equal.py:2079]:
```suggestion
        with pytest.raises(AssertionError, match="The hyperparameter unitary_matrix is not equal"):
```

**Review by lillian542 — APPROVED** [review]
Looks good to me!

**Review by PietropaoloFrisoni — APPROVED** [review]
Thanks!

**Pull request merged**